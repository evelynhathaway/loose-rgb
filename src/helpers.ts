import type {MarkOptional} from "ts-essentials";


/*
	Types
*/
export interface RGBLowerKeys {
	r: number;
	g: number;
	b: number;
}
export interface RGBALowerKeys extends RGBLowerKeys {
	a: number;
}

export interface RGBUpperKeys {
	R: number;
	G: number;
	B: number;
}
export interface RGBAUpperKeys extends RGBUpperKeys {
	A: number;
}

export interface RGBFullKeys {
	red: number;
	green: number;
	blue: number;
}
export interface RGBAFullKeys extends RGBFullKeys {
	alpha: number;
}

export type RGBArray = [number, number, number]
export type RGBAArray = [number, number, number, number]

export type RGBCSSValue = string;
export type RGBACSSValue = string;

export type RGBHex = string;
export type RGBAHex = string;

export type RGB = RGBLowerKeys | RGBUpperKeys | RGBFullKeys | RGBArray;
export type RGBA = RGBALowerKeys | RGBAUpperKeys | RGBAFullKeys | RGBAArray;
export type RGBX = RGB | RGBA;
export type RGBUnknown = RGB | RGBA | RGBCSSValue | RGBACSSValue | RGBHex | RGBAHex;

export type RGBTypeName = "LowerKeys" | "UpperKeys" | "FullKeys" | "Array" | "Hex" | "CSSValue"


/*
	Helpers
*/
// eslint-disable-next-line @typescript-eslint/ban-types
export const includesKeys = (source: Array<string>, target: Object): boolean => (
	source.every((key) => Object.keys(target).includes(key))
);
export const parseDec = (num: string): number => Number.parseInt(num, 10);
export const parseHex = (num: string): number => Number.parseInt(num, 16);


/*
	Regular Expressions
*/
export const regCSSNum = "\\s*(\\d+)\\s*";
export const regexRGBCSSValue = new RegExp(`^rgb\\(${regCSSNum},${regCSSNum},${regCSSNum}\\)$`, "i");
export const regexRGBACSSValue = new RegExp(`^rgba\\(${regCSSNum},${regCSSNum},${regCSSNum},${regCSSNum}\\)$`, "i");

export const regHexNum = "([\\dA-F]{2})";
export const regHexShortNum = "([\\dA-F])";
export const regexRGBHexShort = new RegExp(`^#?${regHexShortNum}${regHexShortNum}${regHexShortNum}$`, "i");
export const regexRGBAHexShort = new RegExp(`^#?${regHexShortNum}${regHexShortNum}${regHexShortNum}${regHexShortNum}$`, "i");
export const regexRGBHex = new RegExp(`^#?${regHexNum}${regHexNum}${regHexNum}$`, "i");
export const regexRGBAHex = new RegExp(`^#?${regHexNum}${regHexNum}${regHexNum}${regHexNum}$`, "i");


/*
	Type Guards
*/
export const isRGBLowerKeys = (input: RGBUnknown): input is MarkOptional<RGBALowerKeys, "a"> => (
	input instanceof Object && includesKeys(["r", "g", "b"], input)
);
export const isRGBUpperKeys = (input: RGBUnknown): input is MarkOptional<RGBAUpperKeys, "A"> => (
	input instanceof Object && includesKeys(["R", "G", "B"], input)
);
export const isRGBFullKeys = (input: RGBUnknown): input is MarkOptional<RGBAFullKeys, "alpha"> => (
	input instanceof Object && includesKeys(["red", "green", "blue"], input)
);
export const isRGBArray = (input: RGBUnknown): input is RGBArray | RGBAArray => (
	Array.isArray(input) && input.length >= 3
);
export const isRGBCSSValue = (input: RGBUnknown): input is RGBCSSValue => (
	typeof input === "string" && regexRGBCSSValue.test(input)
);
export const isRGBACSSValue = (input: RGBUnknown): input is RGBACSSValue => (
	typeof input === "string" && regexRGBACSSValue.test(input)
);
export const isRGBHex = (input: RGBUnknown): input is RGBHex => (
	typeof input === "string" && regexRGBHex.test(input)
);
export const isRGBAHex = (input: RGBUnknown): input is RGBAHex => (
	typeof input === "string" && regexRGBAHex.test(input)
);
export const isRGBHexShort = (input: RGBUnknown): input is RGBHex => (
	typeof input === "string" && regexRGBHexShort.test(input)
);
export const isRGBAHexShort = (input: RGBUnknown): input is RGBAHex => (
	typeof input === "string" && regexRGBAHexShort.test(input)
);
