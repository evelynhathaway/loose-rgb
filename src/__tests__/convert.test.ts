import {convert} from "..";
import {RGBUnknown} from "../helpers";


interface GenericConvert {
	(color: RGBUnknown): RGBUnknown;
}
describe("convert", () => {
	const convertMethods: Array<keyof typeof convert> = [
		"toLowerKeys",
		"toUpperKeys",
		"toFullKeys",
		"toArray",
		"toCSSValue",
		"toHex",
	];
	const convertEdgeCases = [
		{r: 0, g: 1, b: 2},
		{r: 255, g: 101, b: 16},
		{r: 0, g: 1, b: 2, a: 0},
		{r: 255, g: 101, b: 16, a: 0.95},
	];

	describe.each(convertMethods)("%s", (method) => {
		test.each(convertEdgeCases)("%p", (color) => {
			expect((convert[method] as GenericConvert)(color)).toMatchSnapshot();
		});
	});
});
