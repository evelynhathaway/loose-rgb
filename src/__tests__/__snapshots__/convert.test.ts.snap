// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert toArray {"a": 0, "b": 2, "g": 1, "r": 0} 1`] = `
Array [
  0,
  1,
  2,
]
`;

exports[`convert toArray {"a": 0.95, "b": 16, "g": 101, "r": 255} 1`] = `
Array [
  255,
  101,
  16,
  0.95,
]
`;

exports[`convert toArray {"b": 2, "g": 1, "r": 0} 1`] = `
Array [
  0,
  1,
  2,
]
`;

exports[`convert toArray {"b": 16, "g": 101, "r": 255} 1`] = `
Array [
  255,
  101,
  16,
]
`;

exports[`convert toCSSValue {"a": 0, "b": 2, "g": 1, "r": 0} 1`] = `"rgb(0, 1, 2)"`;

exports[`convert toCSSValue {"a": 0.95, "b": 16, "g": 101, "r": 255} 1`] = `"rgba(255, 101, 16, 0.95)"`;

exports[`convert toCSSValue {"b": 2, "g": 1, "r": 0} 1`] = `"rgb(0, 1, 2)"`;

exports[`convert toCSSValue {"b": 16, "g": 101, "r": 255} 1`] = `"rgb(255, 101, 16)"`;

exports[`convert toFullKeys {"a": 0, "b": 2, "g": 1, "r": 0} 1`] = `
Object {
  "blue": 2,
  "green": 1,
  "red": 0,
}
`;

exports[`convert toFullKeys {"a": 0.95, "b": 16, "g": 101, "r": 255} 1`] = `
Object {
  "alpha": 0.95,
  "blue": 16,
  "green": 101,
  "red": 255,
}
`;

exports[`convert toFullKeys {"b": 2, "g": 1, "r": 0} 1`] = `
Object {
  "blue": 2,
  "green": 1,
  "red": 0,
}
`;

exports[`convert toFullKeys {"b": 16, "g": 101, "r": 255} 1`] = `
Object {
  "blue": 16,
  "green": 101,
  "red": 255,
}
`;

exports[`convert toHex {"a": 0, "b": 2, "g": 1, "r": 0} 1`] = `"#00010200"`;

exports[`convert toHex {"a": 0.95, "b": 16, "g": 101, "r": 255} 1`] = `"#ff6510f3"`;

exports[`convert toHex {"b": 2, "g": 1, "r": 0} 1`] = `"#000102"`;

exports[`convert toHex {"b": 16, "g": 101, "r": 255} 1`] = `"#ff6510"`;

exports[`convert toLowerKeys {"a": 0, "b": 2, "g": 1, "r": 0} 1`] = `
Object {
  "b": 2,
  "g": 1,
  "r": 0,
}
`;

exports[`convert toLowerKeys {"a": 0.95, "b": 16, "g": 101, "r": 255} 1`] = `
Object {
  "a": 0.95,
  "b": 16,
  "g": 101,
  "r": 255,
}
`;

exports[`convert toLowerKeys {"b": 2, "g": 1, "r": 0} 1`] = `
Object {
  "b": 2,
  "g": 1,
  "r": 0,
}
`;

exports[`convert toLowerKeys {"b": 16, "g": 101, "r": 255} 1`] = `
Object {
  "b": 16,
  "g": 101,
  "r": 255,
}
`;

exports[`convert toUpperKeys {"a": 0, "b": 2, "g": 1, "r": 0} 1`] = `
Object {
  "B": 2,
  "G": 1,
  "R": 0,
}
`;

exports[`convert toUpperKeys {"a": 0.95, "b": 16, "g": 101, "r": 255} 1`] = `
Object {
  "A": 0.95,
  "B": 16,
  "G": 101,
  "R": 255,
}
`;

exports[`convert toUpperKeys {"b": 2, "g": 1, "r": 0} 1`] = `
Object {
  "B": 2,
  "G": 1,
  "R": 0,
}
`;

exports[`convert toUpperKeys {"b": 16, "g": 101, "r": 255} 1`] = `
Object {
  "B": 16,
  "G": 101,
  "R": 255,
}
`;
